File: src/test/java/im/tox/tox4j/core/callbacks/FriendMessageCallbackTest.java
package im.tox.tox4j.core.callbacks;

import im.tox.tox4j.AliceBobTestBase;
import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.core.ToxCore;
import im.tox.tox4j.core.enums.ToxConnection;
import im.tox.tox4j.core.enums.ToxMessageType;
import im.tox.tox4j.exceptions.ToxException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class FriendMessageCallbackTest extends AliceBobTestBase {

    @NotNull
    @Override
    protected ChatClient newAlice() {
        return new Client();
    }


    private static class Client extends ChatClient {

        public void friendConnectionStatus(final int friendNumber, @NotNull ToxConnection connection) {
            if (connection != ToxConnection.NONE) {
                debug("is now connected to friend " + friendNumber);
                addTask(new Task() {
                    @Override
                    public void perform(@NotNull ToxCore tox) throws ToxException {
                        tox.sendMessage(friendNumber, ToxMessageType.NORMAL, 0, ("My name is " + getName()).getBytes());
                    }
                });
            }
        }

        @Override
        public void friendMessage(int friendNumber, @NotNull ToxMessageType type, int timeDelta, @NotNull byte[] message) {
            debug("received a message: " + new String(message));
            assertEquals(FRIEND_NUMBER, friendNumber);
            assertEquals(ToxMessageType.NORMAL, type);
            assertTrue(timeDelta >= 0);
            assertEquals("My name is " + getFriendName(), new String(message));
            finish();
        }

    }

}

File: src/test/java/im/tox/tox4j/core/callbacks/FriendMessageCallbackTest.scala
package im.tox.tox4j.core.callbacks

import im.tox.tox4j.core.enums.{ ToxConnection, ToxMessageType }
import im.tox.tox4j.testing.autotest.{ AliceBobTest, AliceBobTestBase, ChatClient }
import org.junit.Assert.{ assertEquals, assertTrue }

final class FriendMessageCallbackTest extends AliceBobTest {

  protected override def newAlice(name: String, expectedFriendName: String) = new ChatClient(name, expectedFriendName) {

    override def friendConnectionStatus(friendNumber: Int, connection: ToxConnection): Unit = {
      if (connection != ToxConnection.NONE) {
        debug("is now connected to friend " + friendNumber)
        addTask { tox =>
          tox.sendMessage(friendNumber, ToxMessageType.NORMAL, 0, ("My name is " + selfName).getBytes)
        }
      }
    }

    override def friendMessage(friendNumber: Int, newType: ToxMessageType, timeDelta: Int, message: Array[Byte]): Unit = {
      debug("received a message: " + new String(message))
      assertEquals(AliceBobTestBase.FRIEND_NUMBER, friendNumber)
      assertEquals(ToxMessageType.NORMAL, newType)
      assertTrue(timeDelta >= 0)
      assertEquals("My name is " + expectedFriendName, new String(message))
      finish()
    }
  }

}
